p2p/ipld/plugin/nodes/nodes.go

```diff
@@ -0,0 +1,396 @@
+		"sha2-256-namespace8-flagged",
+		2*namespaceSize+sha256.Size,
+		sumSha256Namespace8Flagged,
+	)
+}
+
+func mustRegisterNamespacedCodec(
+	codec uint64,
+	name string,
+	defaulLength int,
+	hashFunc mh.HashFunc,
+) {
+	if _, ok := mh.Codes[codec]; !ok {
+		// make sure that the Codec wasn't registered from somewhere different than this plugin already:
+		if _, found := mh.Codes[codec]; found {
+			panic(fmt.Sprintf("Codec 0x%X is already present: %v", codec, mh.Codes[codec]))
+		}
+		// add to mh.Codes map first, otherwise mh.RegisterHashFunc would err:
+		mh.Codes[codec] = name
+		mh.Names[name] = codec
+		mh.DefaultLengths[codec] = defaulLength
+
+		if err := mh.RegisterHashFunc(codec, hashFunc); err != nil {
+			panic(fmt.Sprintf("could not register hash function: %v", mh.Codes[codec]))
+		}
+	}
+}
+
+// sumSha256Namespace8Flagged is the mh.HashFunc used to hash leaf and inner nodes.
+// It is registered as a mh.HashFunc in the go-multihash module.
+func sumSha256Namespace8Flagged(data []byte, _length int) ([]byte, error) {
+	isLeafData := data[0] == nmt.LeafPrefix
+	if isLeafData {
+		return nmt.Sha256Namespace8FlaggedLeaf(data[1:]), nil
+	}
+	return nmt.Sha256Namespace8FlaggedInner(data[1:]), nil
+}
+
+var _ plugin.PluginIPLD = &LazyLedgerPlugin{}
+
+type LazyLedgerPlugin struct{}
+
+func (l LazyLedgerPlugin) RegisterBlockDecoders(dec format.BlockDecoder) error {
+	dec.Register(Nmt, NmtNodeParser)
+	return nil
+}
+
+func (l LazyLedgerPlugin) RegisterInputEncParsers(iec coredag.InputEncParsers) error {
+	iec.AddParser("raw", DagParserFormatName, DataSquareRowOrColumnRawInputParser)
+	return nil
+}
+
+func (l LazyLedgerPlugin) Name() string {
+	return "LazyLedger"
+}
+
+func (l LazyLedgerPlugin) Version() string {
+	return "0.0.0"
+}
+
+func (l LazyLedgerPlugin) Init(env *plugin.Environment) error {
+	return nil
+}
+
+// DataSquareRowOrColumnRawInputParser reads the raw shares and extract the IPLD nodes from the NMT tree.
+// Note, to parse without any error the input has to be of the form:
+//
+// <share_0>| ... |<share_numOfShares>
+//
+// To determine the share and the namespace size the constants
+// types.ShareSize and types.NamespaceSize are used.
+func DataSquareRowOrColumnRawInputParser(r io.Reader, _mhType uint64, _mhLen int) ([]node.Node, error) {
+	const extendedSquareSize = 256
+	br := bufio.NewReader(r)
+	nodes := make([]node.Node, 0, extendedSquareSize)
+	nodeCollector := func(hash []byte, children ...[]byte) {
+		cid := mustCidFromNamespacedSha256(hash)
+		switch len(children) {
+		case 1:
+			prependNode(nmtLeafNode{
+				cid:  cid,
+				Data: children[0],
+			}, &nodes)
+		case 2:
+			prependNode(nmtNode{
+				cid: cid,
+				l:   children[0],
+				r:   children[1],
+			}, &nodes)
+		default:
+			panic("expected a binary tree")
+		}
+	}
+	n := nmt.New(
+		sha256.New(),
+		nmt.NamespaceIDSize(namespaceSize),
+		nmt.NodeVisitor(nodeCollector),
+	)
+	for {
+		share := make([]byte, shareSize+namespaceSize)
```

Comment:
Just for readability, I'm not sure `share` is the best name, since it also includes the namespace ID. Maybe `leaf`?

Reply:
Will rename to `namespacedLeaf`, ok?